Automated Smoketest 
Sha: 8f77ea8313832d2f53f3b28b4d4fccfec4901436
Build: failed
Number of Error(s): 11
  Policy CMP0046 is not set: Error on non-existent dependency in
  Policy CMP0046 is not set: Error on non-existent dependency in
/mnt/disk1/home/vamosax/smoketest/PR-8965/HPCC-Platform/esp/bindings/SOAP/soaplib/./../../../../system/security/shared/caching.hpp:76:91: error: invalid use of incomplete type \xe2\x80\x98class CPermissionsCache\xe2\x80\x99
/mnt/disk1/home/vamosax/smoketest/PR-8965/HPCC-Platform/esp/bindings/SOAP/soaplib/./../../../../system/security/shared/caching.hpp:55:32: error: forward declaration of \xe2\x80\x98class CPermissionsCache\xe2\x80\x99
/mnt/disk1/home/vamosax/smoketest/PR-8965/HPCC-Platform/esp/bindings/SOAP/soaplib/./../../../../system/security/shared/caching.hpp:76:124: error: expected \xe2\x80\x98)\xe2\x80\x99 before \xe2\x80\x98;\xe2\x80\x99 token
/mnt/disk1/home/vamosax/smoketest/PR-8965/HPCC-Platform/esp/bindings/SOAP/soaplib/./../../../../system/security/shared/caching.hpp:196:42: error: \xe2\x80\x98timeout\xe2\x80\x99 was not declared in this scope
/mnt/disk1/home/vamosax/smoketest/PR-8965/HPCC-Platform/esp/bindings/SOAP/soaplib/./../../../../system/security/shared/caching.hpp:76:126: error: control reaches end of non-void function [-Werror=return-type]
cc1plus: some warnings being treated as errors
make[2]: *** [esp/bindings/SOAP/soaplib/CMakeFiles/soaplib.dir/__/__/__/platform/espprotocol.cpp.o] Error 1
make[1]: *** [esp/bindings/SOAP/soaplib/CMakeFiles/soaplib.dir/all] Error 2
make: *** [all] Error 2

--- a/system/security/shared/caching.cpp
+++ b/system/security/shared/caching.cpp
@@ -171,11 +171,7 @@ void CResPermissionsCache::add( IArrayOf<ISecResource>& resources )
 //called from within a WriteLockBlock
 void CResPermissionsCache::removeStaleEntries(time_t tstamp)
 {
-    int timeout = m_pParentCache->getCacheTimeout();
-    if(timeout == 0 && m_pParentCache->isTransactionalEnabled())
-        timeout = 10; //Transactional timeout is set to 10 seconds for long transactions that might take over 10 seconds.
-    tstamp -= timeout;
-    if (m_tLastCleanup < tstamp)
+    if (needsCleanup(tstamp, m_pParentCache->getCacheTimeout()))
     {
         MapTimeStamp::iterator i;
         MapTimeStamp::iterator itL    = m_timestampMap.lower_bound(tstamp);
@@ -234,11 +230,24 @@ int CPermissionsCache::lookup( ISecUser& sec_user, IArrayOf<ISecResource>& resou
 
     const char* userId = sec_user.getName();
 
-    //First, clear stale cache entries for this CResPermissionsCache entry
+    //First check if matching cache entry is stale
+    bool needsCleanup = false;
+    {
+        ReadLockBlock readLock(m_resPermCacheRWLock);
+        MapResPermissionsCache::const_iterator i = m_resPermissionsMap.find( userId );
+        if (i != m_resPermissionsMap.end())
+        {
+            CResPermissionsCache* pResPermissionsCache = (*i).second;
+            needsCleanup = pResPermissionsCache->needsCleanup(tstamp, getCacheTimeout());
+        }
+    }
+
+    //clear stale cache entries for this CResPermissionsCache entry
+    if (needsCleanup)
     {
         WriteLockBlock writeLock(m_resPermCacheRWLock);
         MapResPermissionsCache::const_iterator i = m_resPermissionsMap.find( userId );
-        if (i != m_resPermissionsMap.end())
+        if (i != m_resPermissionsMap.end())//Entry could have been deleted by another thread
         {
             CResPermissionsCache* pResPermissionsCache = (*i).second;
             pResPermissionsCache->removeStaleEntries(tstamp);

--- a/system/security/shared/caching.hpp
+++ b/system/security/shared/caching.hpp
@@ -73,6 +73,10 @@ public:
     //removes entries older than tstamp passed in
     //
     virtual void removeStaleEntries(time_t tstamp);
+    virtual bool needsCleanup(time_t now, unsigned timeout)
+    {
+        return m_tLastCleanup < (now - timeout);
+    }
 private:
 
 
@@ -174,11 +178,31 @@ public:
     virtual void add (ISecUser& sec_user);
     virtual void removeFromUserCache(ISecUser& sec_user);
 
-    void  setCacheTimeout(int timeout) { m_cacheTimeout = timeout; }
+#define DEFAULT_CACHE_TIMEOUT_SECONDS 10
+    void  setCacheTimeout(int timeoutSeconds)
+    {
+        m_cacheTimeout = timeoutSeconds;
+        if(m_cacheTimeout == 0 && isTransactionalEnabled())//ensure transactional time is updated
+            setTransactionalCacheTimeout(DEFAULT_CACHE_TIMEOUT_SECONDS); //Transactional timeout is set to 10 seconds for long transactions that might take over 10 seconds.
+        else
+            setTransactionalCacheTimeout(timeoutSeconds);
+    }
     const int getCacheTimeout() { return m_cacheTimeout; }
     bool  isCacheEnabled() { return m_cacheTimeout > 0; }
-    void setTransactionalEnabled(bool enable) { m_transactionalEnabled = enable; }
+
+    void setTransactionalEnabled(bool enable)
+    {
+        m_transactionalEnabled = enable;
+        if(getCacheTimeout() == 0 && enable)//ensure transactional time is updated
+            setTransactionalCacheTimeout(DEFAULT_CACHE_TIMEOUT_SECONDS); //Transactional timeout is set to 10 seconds for long transactions that might take over 10 seconds.
+        else
+            setTransactionalCacheTimeout(getCacheTimeout());
+    }
+    void setTransactionalCacheTimeout(int timeoutSeconds) { m_transactionalCacheTimeout = timeoutSeconds; }
+    const int getTransactionalCacheTimeout() { return m_transactionalCacheTimeout; }
+
     bool isTransactionalEnabled() { return m_transactionalEnabled;}
+
     void flush();
     bool addManagedFileScopes(IArrayOf<ISecResource>& scopes);
     void removeManagedFileScopes(IArrayOf<ISecResource>& scopes);
@@ -196,6 +220,7 @@ private:
 
     int m_cacheTimeout; //cleanup cycle period
     bool m_transactionalEnabled;
+    int m_transactionalCacheTimeout;
 
     MapUserCache m_userCache;
     mutable ReadWriteLock m_userCacheRWLock;    //guards m_userCache

